feat: filter runtime backups by prefix

lenaschoenburg opened the pull request:
This is the first half of https://github.com/camunda/camunda/issues/36162, adding the option to filter backup statuses by prefix.

`GET actuator/backupRuntime` can now take:
- A full backup id, returning the backup status, as before.
- No extra parameter, listing all backup statuses, as before.
- A backup prefix ending in `*`, listing all backup statuses matching that prefix.

The commits should be rather tidy, reviewing them one by one and reading the commit messages is recommended.

lenaschoenburg commented:
Backport to 8.8 is also pending #37374

ChrisKujawa commented on dist/src/main/java/io/camunda/zeebe/shared/management/BackupEndpoint.java:
```diff
@@ -76,7 +76,40 @@ public WebEndpointResponse<?> take(final long backupId) {
+      return new WebEndpointResponse<>(
+          new Error()
+              .message(
+                  "Id '%s' was neither an number nor a prefix ending with '*'"
```
🔧 Following our error guidelines https://github.com/camunda/camunda/wiki/Error-Guidelines

```
"Expected given backup ID to be a number or a prefix ending with '*', but got '%s' which was neither of them."
```

lenaschoenburg replied:
Done in 2c30d35dfc1009c6bdda6caa3e41aeaf36da1909

ChrisKujawa commented on dist/src/test/java/io/camunda/zeebe/shared/management/BackupEndpointStandaloneTest.java:
```diff
@@ -31,14 +31,14 @@ public void shouldCallTakeWhenIsStandalone() {
-    backupEndpointStandalone.list();
-    verify(backupEndpoint).list();
+    backupEndpointStandalone.query(null);
+    verify(backupEndpoint).query(null);
   }
 
   @Test
   public void shouldCallGetWhenIsStandalone() {
-    backupEndpointStandalone.status(11L);
-    verify(backupEndpoint).status(11L);
+    backupEndpointStandalone.query("11");
+    verify(backupEndpoint).query("11");
```
🙃 No given-when-then blocks 😞. Consistency is key.

https://github.com/camunda/camunda/blob/main/docs/testing.md#general-conventions

ChrisKujawa commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/backup/BackupErrorResponseTest.java:
```diff
@@ -126,7 +126,7 @@ void setup() {
     @Test
     void shouldReturn400() {
       // when - then
-      assertThat(backupEndpoint.status(1).getStatus()).isEqualTo(400);
```
❓ This just changed the Java APIs right, but not the actual actuator endpoints/APIs?

lenaschoenburg replied:
yes, the actual REST API didn't change here.

ChrisKujawa commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/api/BackupManager.java:
```diff
@@ -36,13 +36,17 @@ public interface BackupManager {
   ActorFuture<BackupStatus> getBackupStatus(long checkpointId);
 
+  default ActorFuture<Collection<BackupStatus>> listBackups() {
+    return listBackups(null);
```
❓ No Wildcard?

lenaschoenburg replied:
Outdated, this code was overwritten later on.

ChrisKujawa commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/api/BackupManager.java:
```diff
@@ -36,13 +36,17 @@ public interface BackupManager {
    * Get all available backups where status is one of {@link BackupStatusCode#COMPLETED}, {@link
    * BackupStatusCode#FAILED}, {@link BackupStatusCode#IN_PROGRESS}
    *
    * @return a collection of backup status
```
🔧 Please update the java doc

lenaschoenburg replied:
Done in 9f7a80aec1e156f60ab4b7961b9e433d06248df0

ChrisKujawa commented on zeebe/backup/src/test/java/io/camunda/zeebe/backup/management/BackupServiceImplTest.java:
```diff
@@ -417,7 +417,49 @@ void shouldListAvailableBackups() {
+        .thenReturn(CompletableFuture.completedFuture(List.of(backup1, backup2)));
+
+    // when
+    final var backups = backupService.listBackups(partitionId, "*", concurrencyControl).join();
```
❓ Actually here we now us the wildcard, but in the test method we say empty prefix. Do we want to test something different here?

lenaschoenburg replied:
Test was removed again later on anyway.

ChrisKujawa commented on zeebe/protocol/src/main/resources/cluster-management-protocol.xml:
```diff
@@ -57,6 +57,7 @@
     <field name="partitionId" id="1" type="uint16"/>
     <field name="type" id="2" type="BackupRequestType"/>
     <field name="backupId" id="3" type="int64"/>
+    <data name="pattern" id="4" type="varDataEncoding"/>
```
❓ Should some of them be set as optional? 🤔

lenaschoenburg replied:
`varDataEncoding` can't be optional, sadly.

ChrisKujawa commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/api/BackupIdentifierWildcard.java:
```diff
@@ -27,12 +30,128 @@ public interface BackupIdentifierWildcard {
+    prefix.append(wildcard.partitionId().get());
+    prefix.append("/");
+
+    switch (wildcard.checkpointPattern()) {
+      case Exact(final var checkpointId) -> {
+        prefix.append(checkpointId);
+        prefix.append("/");
+        // Checkpoint pattern is exact so we can include node id if present
+        if (wildcard.nodeId().isPresent()) {
+          prefix.append(wildcard.nodeId().get());
+        }
+      }
+      // Checkpoint pattern is not exact so our prefix ends here, we can't include the node id
+      case Prefix(final var cpPrefix) -> prefix.append(cpPrefix);
+      case Any() -> {}
```
💭 Took me a while to understand this. Maybe it is just me, but I feel this could be easier/reduced 🤔

lenaschoenburg replied:
Imagine we add support for matching checkpoints by regex. I think it's way easier to just add one record and then get a compile error telling me where I need to add support, rather than somehow detecting it's a regex by the contents of the string and then going over all usages manually, hoping I don't forget any.

ChrisKujawa commented on zeebe/backup/src/test/java/io/camunda/zeebe/backup/management/BackupServiceImplTest.java:
```diff
@@ -423,48 +424,6 @@ void shouldListAvailableBackups() {
-    // then
-    assertThat(backups).containsExactlyInAnyOrder(backup1, backup2);
-  }
-
```
❓ was this moved?

ChrisKujawa commented on zeebe/backup/src/test/java/io/camunda/zeebe/backup/api/BackupIdentifierWildcardTest.java:
```diff
@@ -0,0 +1,184 @@
+  void shouldBuildPrefixWithPartitionId() {
+    // given
+    final var wildcard =
+        new BackupIdentifierWildcardImpl(Optional.empty(), Optional.of(1), new Any());
```
❓ Wondering about the ordering of the parameters. Wouldn't it make more sense to have the same order here as the output?

Like first partition id, then checkpoint and at the end the node id?

ChrisKujawa commented on zeebe/backup/src/test/java/io/camunda/zeebe/backup/api/BackupIdentifierWildcardTest.java:
```diff
@@ -0,0 +1,184 @@
+    // given
+    final var pattern = new Exact(100);
+    // then
+    assertThat(pattern.matches(100L)).isTrue();
```
❓ What about `1000`?

ChrisKujawa commented on zeebe/backup/src/test/java/io/camunda/zeebe/backup/api/BackupIdentifierWildcardTest.java:
```diff
@@ -0,0 +1,184 @@
+    final var pattern = new Prefix("10");
+    // then
+    assertThat(pattern.matches(10L)).isTrue();
+    assertThat(pattern.matches(100L)).isTrue();
```
❓ what about `110`?

lenaschoenburg replied:
Will add that case 👍

ChrisKujawa commented on zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/BackupActuator.java:
```diff
@@ -112,6 +112,10 @@ static BackupActuator of(final String endpoint) {
 
+  @RequestLine("GET /{prefix}")
+  @Headers({"Content-Type: application/json", "Accept: application/json"})
+  List<BackupInfo> list(@Param final String prefix);
```
❓ The other list will always do with wildcard right?

lenaschoenburg replied:
Exactly :+1: That's how the API works now.

ChrisKujawa reviewed (approved):
Thanks @lenaschoenburg it was easy to follow 👍🏼, while I was wondering about this, as I thought we already had this implemented. Was this maybe just for the backupHistory? 

The general changes make sense to me, and I see no blocker.
I simply have some questions/comments to clarify and to potentially improve. Please look at them before merging. Thanks 🙇🏼

lenaschoenburg commented:
> I was wondering about this, as I thought we already had this implemented. Was this maybe just for the backupHistory?

That's right, this was the issue: https://github.com/camunda/camunda/issues/30694

I think I addressed all suggestions and questions, merging :tada:

backport-action commented:
Created backport PR for `stable/8.6`:
- #37572 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-37460-to-stable/8.6
git worktree add --checkout .worktree/backport-37460-to-stable/8.6 backport-37460-to-stable/8.6
cd .worktree/backport-37460-to-stable/8.6
git reset --hard HEAD^
git cherry-pick -x 87894c58280c4460d4de73be6a586078f4509b2c e9e4cb9cedc2b2bc10af88a48b5a568b8adb8415 be2f86525d26bf2864af7ad86a49d58c27edf96a e08d8fcbae341c5d3641b9a7df4abf633d072945 fb4d292b8d48369d54552b04025d36dc8c744b05 fa81b7b8bf14d0851c9c689c086379e108c5cc02 2c30d35dfc1009c6bdda6caa3e41aeaf36da1909 429e22cb1bb4bb28e7e1233ab234d64030c1caa7 2a4292cce73c738a78d640ffd9caf552d811624a 9f7a80aec1e156f60ab4b7961b9e433d06248df0 f7a9a4da64d95f6db02c11a2238fe9afad868d28
git push --force-with-lease
```

backport-action commented:
Created backport PR for `stable/8.7`:
- #37573 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-37460-to-stable/8.7
git worktree add --checkout .worktree/backport-37460-to-stable/8.7 backport-37460-to-stable/8.7
cd .worktree/backport-37460-to-stable/8.7
git reset --hard HEAD^
git cherry-pick -x 87894c58280c4460d4de73be6a586078f4509b2c e9e4cb9cedc2b2bc10af88a48b5a568b8adb8415 be2f86525d26bf2864af7ad86a49d58c27edf96a e08d8fcbae341c5d3641b9a7df4abf633d072945 fb4d292b8d48369d54552b04025d36dc8c744b05 fa81b7b8bf14d0851c9c689c086379e108c5cc02 2c30d35dfc1009c6bdda6caa3e41aeaf36da1909 429e22cb1bb4bb28e7e1233ab234d64030c1caa7 2a4292cce73c738a78d640ffd9caf552d811624a 9f7a80aec1e156f60ab4b7961b9e433d06248df0 f7a9a4da64d95f6db02c11a2238fe9afad868d28
git push --force-with-lease
```

backport-action commented:
Successfully created backport PR for `stable/8.8`:
- #37574
